WordPress/wordpress-playground
# CORS headers for client/index.js

## What is this PR doing?

This allows any domain to `import()` the `client/index.js` file from https://playground.wordpress.net/.

It restores the changes described in:
https://github.com/WordPress/wordpress-playground/issues/873

More specifically, https://github.com/WordPress/wordpress-playground/pull/773 exposed the client library on Playground.wordpress.net:

https://playground.wordpress.net/client/index.js

That PR was accompanied by a .htaccess update to ensure the correct Access-Control-Allow-Origin header was set – otherwise the index.js file cannot be import()-ed in JavaScript.

Unfortunately, that update was lost in a force-push. I patched it ad-hoc on the server, but we should ship the correct setup out of the box.

Luckily, GitHub still keeps that code change here:

https://github.com/WordPress/wordpress-playground/commit/a6f685d0ac154153c00b421adb24e9a161ac77b6

## What problem is it solving?

Without these changes the `index.js` file cannot be `import()`-ed in JavaScript when pulling the file from a different domain. This adds the `set Access-Control-Allow-Origin "*"` header that allow JS running on any domain to `import()` that file.

## How is the problem addressed?

Other domains aren't able to `import()` the `client/index.js` file from https://playground.wordpress.net/ without CORS headers. Right now, the `.htaccess` file has been placed ad-hoc, which is a fragile change that is prone to being lost. 

This codifies the publishing of the `.htaccess` file as a part of the build process, so it won't be lost or overwritten accidentally.

## Testing Instructions

Checkout the branch and install deps:

```bash
git checkout sm-cors-headers-client-js
npm install
```

Run the build:

```bash
npx nx reset
npm run build
```

Ensure the following files exist & match the contents below:

`dist/packages/playground/remote/.htaccess`
```
AddType application/wasm .wasm
AddType	application/octet-stream .data
```

`dist/packages/playground/website/.htaccess`:
```
<FilesMatch "(client\/|blueprint-schema\.json)">
Header set Access-Control-Allow-Origin "*"
</FilesMatch>
```

`dist/packages/playground/wasm-wordpress-net/.htaccess`:
```
AddType application/wasm .wasm
AddType	application/octet-stream .data
<FilesMatch "(client\/|blueprint-schema\.json)">
Header set Access-Control-Allow-Origin "*"
</FilesMatch>
```
